feat: Add support for injecting Debug IDs

timfish opened the pull request:
This PR contains:

- [x] feature

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)

Breaking Changes?

- [x] no

### Description

Adds support for injecting debug IDs into source files and sourcemaps as per the [TC39 proposal](https://github.com/tc39/source-map/blob/main/proposals/debug-id.md).

- Sentry users have been relying on debug ids in production for a number of years. Sentry process hundreds of millions of files per month containing debug ids!
- [A Github search](https://github.com/search?q=%2F%5C%2F%5C%2F%23%20debugId%3D%2F&type=code) suggests debug ids are also used by Backtrace and Expo
- Bun added debug id injection to it's bundler when they added sourcemap support
- Rolldown [just merged debug id support](https://github.com/rolldown/rolldown/pull/2516)

Imran-imtiaz48 reviewed (commented):
The recent changes introduce new flexibility and configurability to the handling of sourcemaps and output validation, which is a welcome improvement. Here are some thoughts and observations on the adjustments:

1. **sourcemapFileNames Enhancements**: 
   - The added flexibility in defining `sourcemapFileNames` either as a `string` or a function (`(chunkInfo: PreRenderedChunk) => string`) is a good approach. This allows for more customized and dynamic sourcemap file naming. Consider documenting examples of how this can be used to help other developers understand the potential benefits of this feature.

2. **sourcemapIgnoreList Option**: 
   - The introduction of `sourcemapIgnoreList` is great for situations where you want to ignore specific files from sourcemapping. If possible, it might be helpful to expand on how `SourcemapIgnoreListOption` works. Are there specific cases or examples that could be highlighted to guide its usage?

3. **sourcemapPathTransform Functionality**:
   - Including `sourcemapPathTransform` allows for powerful path transformations, which will be especially helpful for complex build environments. Clarifying its usage and providing a brief note on how it transforms paths could prevent confusion.

4. **Debugging with sourcemapDebugIds**:
   - The `sourcemapDebugIds` flag is an interesting addition. It would be beneficial to explain the scenarios where enabling this would be most helpful. This could enhance debugging processes, but highlighting its impact on performance or build complexity would be useful.

5. **strict and systemNullSetters**:
   - Keeping the `strict` and `systemNullSetters` options ensures robust validation and system behavior. However, it might be worthwhile to indicate any potential performance impacts or if there are edge cases that developers should be aware of when using these settings.

6. **Overall Validation**:
   - The `validate` flag is crucial for maintaining the integrity of the output. It’s always reassuring to see validation options prominently available to avoid runtime issues.

### Recommendations:
- **Documentation**: Consider enhancing the documentation with examples and scenarios for these settings. This will make the new features more approachable and usable for a broader range of developers.
- **Performance Considerations**: If any of these settings significantly impact build or runtime performance, detailing this in the documentation or with warning notes could be helpful.

timfish commented:
>  It would be beneficial to explain the scenarios where enabling this would be most helpful. 

It's mainly to help match up sourcemaps to their respective source files when de-minifying production stack traces. Without debug ids all you have to find the correct sourcemap is the `sourceMappingURL`. If your app has release versions you can use that and the `sourceMappingURL` to find the correct sourcemap but this has many pitfalls. 

Which release version are you using? What if your app is made up from many different components which are all released individually. What if the `sourceMappingURL` is wrong? Getting the correct `sourceMappingURL` emitted into a sourcemaps has plagued devs for years. 

If both source and sourcemaps contain a matching unique id, it's very easy to match them up and the above issues disappear.

This is not a new idea. Native builds and debug symbols have been doing this for decades!


> but highlighting its impact on performance or build complexity would be useful.

Rolldown always computes hashes for every file so injecting debug ids has minimal performance impact.

lukastaegert reviewed (commented):
This looks actually good to me, thank you so much for taking the time!

What is missing to merge this would be updating documentation. I can support you, but you might be able to write a better short description of the option. These are the places:

- `docs/configuration-options/index.md`: I would add an entry to `Advanced functionality`. As a boolean, it would be similar to e.g. `output.inlineDynamicImports`. Note that options are sorted alphabetically here within each section.
- The option should be added to the list in `docs/javascript-api/index.md` in the `### outputOptions object` section. Again, note the alphabetical sorting.
- Then it should be added to the `## Command line flags` section in `docs/command-line-interface/index.md`. Again, note the sorting. Each line has a short description, but it should observe the 80 character limit for each line.
- Last, this section is copied over into `cli/help.md`, which is the source for the command line help.

Thank you and let me know if you need help.

timfish commented:
From the looks of the coverage failure, I need to add a test for debug ids + files with placeholders

lukastaegert commented:
> From the looks of the coverage failure, I need to add a test for debug ids + files with placeholders?

Yes. To do that, you could set `options.output.entryFileNames` to something like `[hash].js` in your test `_config.js` file. Basically it means that file has a hash in its file name.

timfish commented on docs/configuration-options/index.md:
```diff
@@ -1744,6 +1744,16 @@ If `true`, a separate sourcemap file will be created. If `"inline"`, the sourcem
+|     CLI: | `--sourcemapDebugIds`/`--no-sourcemapDebugIds` |
+| Default: | `false`                                        |
+
+if `true`, unique ids will be emitted in source and sourcemaps which streamlines identifying sourcemaps across different builds.
```
I originally added a link but removed it in favour of not adding something that adds extra maintenance and needS changing later. 

I can add this back but we'll need to make a note somewhere to remember to update it!

lukastaegert replied:
I think linking to the proposal at least will not become "wrong", and it will give interested users some context.

timfish commented:
I think this PR is now ready.

- Added another test to give full test coverage
- Added a link to the TC39 proposal
- Fixed the missing arg in `help.md`

lukastaegert reviewed (approved):
Looks really good!
